### PR TITLE
Do not lock on reads of XPackLicenseState

### PR DESCRIPTION
**Pull request by Tim-Brooks**
XPackLicenseState reads to necessary to validate a number of cluster
operations. This reads occasionally occur on transport threads which
should not be blocked. Currently we sychronize when reading. However,
this is unecessary as only a single piece of state is updateable. This
commit makes this state volatile and removes the locking.

### CODE REVIEW

**Comment by tbrooks8** [comment]
@jasontedor @jaymode - This back port required a number of additional changes, please someone review to spot check for errors.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]:
```suggestion
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]:
```suggestion
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]:
```suggestion
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]:
```suggestion
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]:
```suggestion
```

**Review by jaymode — APPROVED** [review]
LGTM

**Comment by ywangd** [comment]
@tbrooks8 I merged some change to this file in master and your one was on top it. But I haven’t backported my changes yet. The PR !52115  is ready, but my intentions was to bundle the it with a few more refactoring. Not sure if you had to solve many conflicts with your backporting. I wonder whether it would be better to have my backport go through first to maintain consistency with original work? 
@tvernum what do you think?

**Comment by tbrooks8** [comment]
@ywangd - I will wait until you merge.

**Comment by ywangd** [comment]
Thanks. I'll get it sorted out asap

**Comment by ywangd** [comment]
@tbrooks8 I have merged my backport. Thanks a lot for coordinating on this.

**Comment by ywangd** [comment]
@tbrooks8 Merge of #52115 caused some code conflicts for your PR. I am happy to assist in resovling them. Please let me know how I can help. Thanks.

**Comment by tbrooks8** [comment]
@ywangd  I just cherry-picked from master again and it was clean this time.

**Pull request merged**
